### PR TITLE
Remove last connect semantics from reset inference

### PR DESCRIPTION
**Pull request by jackkoenig**
**TLDR**
* Reverts support for last connect semantics in reset inference (reverts https://github.com/freechipsproject/firrtl/pull/1291, note this was never released)
* Makes reset inference a forward and backward propagation
    * This mainly catches user error in InferResets instead of CheckTypes and can change the type of wires or ports that are **only** invalidated but drive stuff (not necessarily the winning connection though). Note that this case crashes in `1.2.2`.
* Fixes bugs
    * Handling of invalidation on components of type `Reset`
    * Connections or invalidation of submodule ports could override everything done inside the module itself
    * Support AsyncResetType in RemoveValidIfs
* This is orthogonal from possible improvements to the feature like https://github.com/freechipsproject/firrtl/issues/1376. This makes the `1.2.x` branch work in a reasonable way.

**Note** this looks big, but the positive line change is almost entirely new tests

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?

### Type of Improvement

- bug fix
- new feature/API

### API Impact

This modifies the reset inference API. It reverts support for last connect semantics in inference. This made it to `1.2.x`, but has not yet been released.

This also fixes many bugs present in the current latest release `1.2.2`, especially with regards to invalidation and inference. It does make a change from `1.2.2`, namely that invalidated components that drive another component that itself is driven by something else may get a different type. **While this may seem like cause for concern, as far as I can tell, every case of this crashes on v1.2.2**; not to mention that such components are only invalidated so their types don't really have much impact on anything.

I think the best way to understand this PR is with the new tests:

1. 
```scala
  it should "infer based on what a component *drives* not just what drives it" in {
    val result = compile(s"""
      |circuit top :
      |  module top :
      |    input in : AsyncReset
      |    output out : Reset
      |    wire w : Reset
      |    w is invalid
      |    out <= w
      |    out <= in
      |""".stripMargin)
    result should containTree { case DefWire(_, "w", AsyncResetType) => true }
  }
```
This first one above is the "API change" that technically would differ from `1.2.2`, but it crashes in `1.2.2`. My argument is that it's not an API that means anything because it only matters in the context of components that are **only ever invalidated**, thus the reset should be optimized away anyway, and if optimizations are disabled, the flop will never be reset.

2.
```scala
  it should "infer from connections, ignoring \"winning\" invalidation" in {
    val result = compile(s"""
      |circuit top :
      |  module top :
      |    input in : AsyncReset
      |    output out : Reset
      |    out <= in
      |    out is invalid
      |""".stripMargin)
    result should containTree { case Port(_, "out", Output, AsyncResetType) => true }
  }
```
With the unreleased last connect semantics, `out` would be a `UInt<1>` instead of `AsyncReset`. Note that as above, it is invalidated so it doesn't do anything. Also note that this crashes in `1.2.2`.

3. 
```scala
  // The backwards type propagation constrains `w` to be the same as both `out0` and `out1`
  it should "not allow an invalidated Wire to drive both a UInt<1> and an AsyncReset" in {
    an [InferResets.InferResetsException] shouldBe thrownBy {
      val result = compile(s"""
        |circuit top :
        |  module top :
        |    input in0 : AsyncReset
        |    input in1 : UInt<1>
        |    output out0 : Reset
        |    output out1 : Reset
        |    wire w : Reset
        |    w is invalid
        |    out0 <= w
        |    out1 <= w
        |    out0 <= in0
        |    out1 <= in1
        |""".stripMargin
      )
    }
  }
```
This also crashes on `v1.2.2`. It would be legal if last connect semantics were accepted, but due to backwards type inference introduced in this PR, `w` is not allowed to drive both a `UInt<1>` and an `AsyncReset`.

4. 

```scala
  // This tests for a bug unrelated to support or lackthereof for last connect in inference
  it should "take into account both internal and external constraints on Module port types" in {
    val result = compile(s"""
      |circuit top :
      |  module child :
      |    input i : AsyncReset
      |    output o : Reset
      |    o <= i
      |  module top :
      |    input in : AsyncReset
      |    output out : AsyncReset
      |    inst c of child
      |    c.o is invalid
      |    c.i <= in
      |    out <= c.o
      |""".stripMargin)
    result should containTree { case Port(_, "o", Output, AsyncResetType) => true }
  }
```
Unrelated bugfix that can lead to very strange behavior.

5.

```scala
  it should "not crash on combinational loops" in {
    an [CheckCombLoops.CombLoopException] shouldBe thrownBy {
      val result = compile(s"""
        |circuit top :
        |  module top :
        |    input in : AsyncReset
        |    output out : Reset
        |    wire w0 : Reset
        |    wire w1 : Reset
        |    w0 <= in
        |    w0 <= w1
        |    w1 <= w0
        |    out <= in
        |""".stripMargin,
        compiler = new LowFirrtlCompiler
      )
    }
  }
```
This crashes on `1.2.2` and `1.2.x` alike.

### Backend Code Generation Impact

It could change the types of some registers, but only if the reset signal **is invalid**.

### Desired Merge Strategy

- Squash: The PR will be squashed and merged

The history is ugly but I want it on the PR to keep some context. This should be squashed so that it doesn't show up in the actual git history.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [x] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/test/scala/firrtlTests/InferResetsSpec.scala [hunk 1]
```diff
@@ -410,5 +405,90 @@ class InferResetsSpec extends FirrtlFlatSpec {
...
+      |    output out : Reset
+      |    wire w : Reset
+      |    w is invalid
+      |    out <= w
+      |    out <= in
+      |""".stripMargin)
+    result should containTree { case DefWire(_, "w", AsyncResetType) => true }
+  }
+
+  it should "infer from connections, ignoring \"winning\" invalidation" in {
```
Comment (albert-magyar) [inline src/test/scala/firrtlTests/InferResetsSpec.scala]: Maybe reword this as "ignoring the fact that the invalidation wins"?

Comment (jackkoenig) [reply src/test/scala/firrtlTests/InferResetsSpec.scala]: yeah that's better

File: src/test/scala/firrtlTests/InferResetsSpec.scala [hunk 2]
```diff
@@ -410,5 +405,90 @@ class InferResetsSpec extends FirrtlFlatSpec {
...
+      |    inst c of child
+      |    c.o is invalid
+      |    c.i <= in
+      |    out <= c.o
+      |""".stripMargin)
+    result should containTree { case Port(_, "o", Output, AsyncResetType) => true }
+  }
+
+  it should "not crash on combinational loops" in {
+    an [CheckCombLoops.CombLoopException] shouldBe thrownBy {
```
Comment (seldridge) [inline src/test/scala/firrtlTests/InferResetsSpec.scala]:
```suggestion
    a [CheckCombLoops.CombLoopException] shouldBe thrownBy {
```

File: src/test/scala/firrtlTests/InferResetsSpec.scala [hunk 3]
```diff
@@ -410,5 +405,109 @@ class InferResetsSpec extends FirrtlFlatSpec {
...
+      |    output out : Reset
+      |    wire w : Reset
+      |    w is invalid
+      |    out <= w
+      |    out <= in
+      |""".stripMargin)
+    result should containTree { case DefWire(_, "w", AsyncResetType) => true }
+  }
+
+  it should "infer from connections, ignoring \"winning\" invalidation" in {
```
Comment (albert-magyar) [inline src/test/scala/firrtlTests/InferResetsSpec.scala]:
```suggestion
  it should "infer from connections, ignoring the fact that the invalidation wins" in {
```

File: src/main/scala/firrtl/transforms/InferResets.scala [hunk 4]
```diff
@@ -156,54 +177,53 @@ class InferResets extends Transform {
...
+          graph.addPairWithEdge(v, u)
+          graph.addPairWithEdge(u, v)
+        case TypeDriver(tpe, _) =>
+          // Use nodes we already made, saves memory
+          val u = tpe match {
+            case AsyncResetType => asyncNode
+            case Utils.BoolType => syncNode
+            case other          => throwInternalError(s"Shouldn't have $other here")
+          }
+          graph.addPairWithEdge(u, v)
```
Comment (albert-magyar) [inline src/main/scala/firrtl/transforms/InferResets.scala]: This "fake" edge to the type seems like a clever way to share the work for the inference with the check _without_ interspersing check code into the connection edge creation.

File: src/main/scala/firrtl/transforms/InferResets.scala [hunk 5]
```diff
@@ -156,54 +177,53 @@ class InferResets extends Transform {
...
+          graph.addPairWithEdge(v, u)
+        case InvalidDriver =>
+          graph.addVertex(v)   // Must be in the graph or won't be inferred
+      }
     }
+    val async = graph.reachableFrom(asyncNode)
+    val sync = graph.getVertices -- async
```
[the remark this answers is not in the document]
Comment (jackkoenig) [reply src/main/scala/firrtl/transforms/InferResets.scala]: I'll write a comment how it works because it may be non-obvious in a future fix/refactor.

File: src/main/scala/firrtl/transforms/InferResets.scala [hunk 6]
```diff
@@ -156,54 +177,53 @@ class InferResets extends Transform {
...
     Try {
-      errors.trigger()
-      res.toMap
+      if (async.contains(syncNode)) throw InferResetsException(graph.path(asyncNode, syncNode))
+      ( sync.view.collect { case Var(t) => t -> syncNode.tpe } ++
+       async.view.collect { case Var(t) => t -> asyncNode.tpe }).toMap
```
[the remark this answers is not in the document]
Comment (jackkoenig) [reply src/main/scala/firrtl/transforms/InferResets.scala]: I like that better, I'll do that

**Comment by jackkoenig** [comment]
Once this passes tests, I'm gonna :shipit: (but will squash and merge it myself to cleanup commit message)
